### PR TITLE
add provider support for inserting new field (after existing fields) to existing model

### PR DESCRIPTION
**Pull request by marcardar**
The purpose of this PR is to support the case when adding a new template because that template may need a corresponding conditional field. The new field is appended to the end of the existing fields so will not conflict with the position (or existence) of existing fields.

### CODE REVIEW

**Comment by marcardar** [comment]
Fixed bug where Models.flush() was missing (also in previous PR). Easy mistake to make, eh!?

**Comment by marcardar** [comment]
Might be nice to support a default value too (set all notes to that value for that new field). So that all existing notes could be enabled...

**Comment by timrae** [comment]
Could you add a unit test please?

**Comment by marcardar** [comment]
Added unit test. Please check.

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 1]
```diff
@@ -949,15 +951,40 @@ public Uri insert(Uri uri, ContentValues values) {
                         throw new RuntimeException(e);
                     }
                     models.addTemplate(existingModel, t);
-                    col.getModels().save(existingModel);
+                    models.save(existingModel);
+                    models.flush();
```
Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: Tim - please could you look through this class and check whether other models.flush() calls need to be made?  I don't understand the libanki behaviour well enough to make that call.

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: I don't really have time, but I assume all of the model / template insert and update methods need it...? Maybe we could add a `col.close(false)` call to the unit tests (followed by a re-open) to check that the changes have been flushed to the db?

**Comment by marcardar** [comment]
Thanks! I was expecting a lengthy discussion :)

I think it's important not to forget two outstanding points here:

1. Are there any more missing Modes.flush() calls?
2. Add support for a default value for the new field?

**Comment by timrae** [comment]
> Are there any more missing Modes.flush() calls?

See above.

> Add support for a default value for the new field?

I don't consider this a good feature for the content provider. In general I don't envisage the provider supporting anything that's not already in libanki.

**Comment by marcardar** [comment]
> In general I don't envisage the provider supporting anything that's not already in libanki.

Okay. So, it's really a LibAnkiProvider. It might be worth putting something about this in the class javadoc to make it clear to future devs wanting to make changes.

**Comment by timrae** [comment]
Yeah it's meant to be a wrapper around libanki and the database in general for some performance critical tasks. Feel free to add javadoc

**Comment by marcardar** [comment]
This goes back to some thoughts I mentioned earlier. A LibAnkiProvider makes sense to support other Anki clients (like AnkiWear) but not so much for an AddContentProvider like for dictionary apps.

An AddContentProvider would only query/insert/update fields the client knows about (and in the correct order). Same for tags.  It's a much safer solution than expecting the client app to do all the field mapping and tag handling.

**Pull request merged**